zds/tutorialv2/tests/tests_models.py

```diff
@@ -424,6 +429,34 @@ def test_ensure_slug_stay(self):
         for index, child in enumerate(current.children[0].children):
             self.assertEqual(child.slug, chapter.children[index].slug)  # slug remains for extract as well !
 
+    def test_publication_and_attributes_consistency(self):
+        article = PublishedContentFactory(type="ARTILCE", author_list=[self.user_author])
+        old_date = article.publication_date
+        old_title = article.public_version.title
+        old_description = article.public_version.description
+        self.assertEqual(
+            self.client.login(
+                username=self.user_author.username,
+                password='hostel77'),
+            True)
+        time.sleep(2)
```

Comment:
Seul problème :

>>> from datetime import datetime
>>> datetime.now = lambda ():1
  File "<stdin>", line 1
    datetime.now = lambda ():1
                           ^
SyntaxError: invalid syntax
>>> datetime.now = lambda :1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't set attributes of built-in/extension type
'datetime.datetime'


Mais c'est l'idée de base.

Le 16 octobre 2015 11:39, Quentin Pradet <notifications@github.com> a écrit
:

> In zds/tutorialv2/tests/tests_models.py
> <https://github.com/zestedesavoir/zds-site/pull/3091#discussion_r42225016>
> :
>
> > @@ -424,6 +429,34 @@ def test_ensure_slug_stay(self):
> >          for index, child in enumerate(current.children[0].children):
> >              self.assertEqual(child.slug, chapter.children[index].slug)  # slug remains for extract as well !
> >
> > +    def test_publication_and_attributes_consistency(self):
> > +        article = PublishedContentFactory(type="ARTILCE", author_list=[self.user_author])
> > +        old_date = article.publication_date
> > +        old_title = article.public_version.title
> > +        old_description = article.public_version.description
> > +        self.assertEqual(
> > +            self.client.login(
> > +                username=self.user_author.username,
> > +                password='hostel77'),
> > +            True)
> > +        time.sleep(2)
>
> @pierre-24 <https://github.com/pierre-24> Ah, je me suis encore fait
> avoir par l'utilisation de belge de "savoir" ! Désolé.
>
> Regarde un exemple avec unittest.mock()
> <https://docs.python.org/3/library/unittest.mock.html#quick-guide%20par%20exemple>.
> L'idée ici est de remplacer datetime.now() pour avoir deux secondes de
> retard avant de republier, ce qui évite d'attendre deux vraies secondes.
> (Pas besoin de se rappeler du nombre d'appels je pense.)
>
> Le fait de changer datetime.now() est très couplé à l'implémentation de
> publish_content
> <https://github.com/artragis/zds-site/blob/use_published_content/zds/tutorialv2/utils.py#L664>.
> On pourrait aussi avoir une fonction intermédiaire pour mocker celle-là au
> lieu de datetime.now(). Mais c'est pas très grave : l'important est
> d'éviter la catastrophe qui est d'attendre deux secondes pour un seul test.
> Deux secondes, c'est le max que devraient prendre des bons tests pour
> s'exécuter, ce qui permet d'avoir du feedback en boucle et d'aller plus
> vite.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zestedesavoir/zds-site/pull/3091/files#r42225016>.
>